hazelcast/src/main/java/com/hazelcast/map/impl/query/MapLocalParallerQueryRunner.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.query;
+
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.map.impl.MapServiceContext;
+import com.hazelcast.query.PagingPredicate;
+import com.hazelcast.query.Predicate;
+import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.query.impl.predicates.QueryOptimizer;
+import com.hazelcast.util.executor.ManagedExecutorService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.map.impl.query.MapQueryEngineUtils.waitForResult;
+import static com.hazelcast.query.PagingPredicateAccessor.getNearestAnchorEntry;
+import static com.hazelcast.util.SortingUtil.getSortedSubList;
+
+/**
+ * Specialization of the {@link MapLocalQueryRunner} - the only difference is that
+ * the query evaluation per partition is run in a PARALLEL fashion.
+ *
+ * Runs query operations in the calling thread (thus blucking it)
+ * Query evaluation per partition is run in a PARALLEL fashion.
+ * <p>
+ * Used by query operations only: QueryOperation & QueryPartitionOperation
+ * Should not be used by proxies or any other query related objects.
+ */
+public class MapLocalParallerQueryRunner extends MapLocalQueryRunner {
```

Comment:
I agree there's two concerns there but I think you slightly missed the point by saying `This class takes care of assigning a thread to the actual query`. The runner does not assign a thread to run a query. The query consists of a couple of phases:

- run index query
- run full-partition scan

In the MapLocalQueryRunner all phases are run within the calling thread.
In the MapLocalParallelQueryRunner most of the stuff hapens in the calling thread, apart from the full-partitionScan predicate evaluation - this is delegated to the executor. However, this is encapsulated and the caller does not know about it since the call is blocking anyway.

In order to do what you are suggesting I would have to create the following classes and then use these classes from within the LocalQueryRunner.

- PartitionScanRunner (just have the logic of partition scanning)
- PartitionScanExecutor (runs the partitions can logic - either in the same thread or in the executor)

This comment is appreciated and I see your point, but it looks a bit like an overkill for now since there's only two implementations and their code is very short. 
Till now everything happened in one class, so I guess we can postpone this discussion and create even more fine-grained design on the way.